src/Microsoft.AspNetCore.Http.Connections/Internal/TimerAwaitable.cs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics.Tracing;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Http.Connections.Internal
+{
+    public class TimerAwaitable : IDisposable, ICriticalNotifyCompletion
+    {
+        private Timer _timer;
+        private Action _callback;
+        private static readonly Action _callbackCompleted = () => { };
+
+        private TimeSpan _period;
+
+        private TimeSpan _dueTime;
+        private bool _disposed;
+        private volatile bool _running = true;
+
+        public TimerAwaitable(TimeSpan dueTime, TimeSpan period)
+        {
+            _dueTime = dueTime;
+            _period = period;
+        }
+
+        public bool IsRunning => _running;
+
+        public void Start()
+        {
+            if (_timer == null)
+            {
+                lock (this)
+                {
+                    if (_disposed)
+                    {
+                        return;
+                    }
+
+                    if (_timer == null)
+                    {
+                        _timer = new Timer(state => ((TimerAwaitable)state).Tick(), this, _dueTime, _period);
+                    }
+                }
+            }
+        }
+
+        public TimerAwaitable GetAwaiter() => this;
+        public bool IsCompleted => ReferenceEquals(_callback, _callbackCompleted);
+
+        public void GetResult()
+        {
+            _callback = null;
+        }
+
+        private void Tick()
+        {
+            var continuation = Interlocked.Exchange(ref _callback, _callbackCompleted);
+            continuation?.Invoke();
+        }
+
+        public void OnCompleted(Action continuation)
+        {
+            if (ReferenceEquals(_callback, _callbackCompleted) ||
+                ReferenceEquals(Interlocked.CompareExchange(ref _callback, continuation, null), _callbackCompleted))
+            {
+                Task.Run(continuation);
+            }
+        }
+
+        public void UnsafeOnCompleted(Action continuation)
+        {
+            OnCompleted(continuation);
+        }
+
+        public void Stop()
+        {
+            _running = false;
+
+            Tick();
```

Comment:
You're going to fire the timer when someone calls `Stop`?

Reply:
You have to yield the await.

Reply:
Should it return a bool then so you can skip your action? Then the loop is `while(await _nextHeartbeat) { ... }`

I get that you're obsessed with event loops now but I'm not 100% convinced that this is "better" yet...

Reply:
Seems fine, it's 1000% better objectively so but to be honest, I don't really care if we merge this change right now. It's late. I can do it in another release. But:
- It removes the locking
- It's pull so there's no overlapping ticks possible
- It fixes the races with disposal and changing the timer interval
- The code that runs in the tick callback can be async without being async void

Reply:
Ok, I'm convinced, let's do it